domain/externalcontroller/service/service.go

```diff
@@ -0,0 +1,38 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/crossmodel"
+)
+
+// State describes retrieval and persistence methods for storage.
+type State interface {
+	// UpdateExternalController persists the input controller
+	// record and associates it with the input model UUIDs.
+	UpdateExternalController(ctx context.Context, ec crossmodel.ControllerInfo, modelUUIDs []string) error
+}
+
+// Service provides the API for working with external controllers.
+type Service struct {
+	st State
+}
+
+// NewService returns a new service reference wrapping the input state.
+func NewService(st State) *Service {
+	return &Service{st}
+}
+
+// UpdateExternalController persists the input controller
+// record and associates it with the input model UUIDs.
+func (s *Service) UpdateExternalController(
+	ctx context.Context, ec crossmodel.ControllerInfo, modelUUIDs ...string,
```

Comment:
It would be really nice to validate the modelUUIDs as UUIDs. I'd rather we didn't keep passing strings around all the time, but types that actually mean something even if it's just `type UUID string`.

Reply:
I will see what we can do here when we roll into the API server.